src/rendezvous/channels/MSC3903ECDHv1RendezvousChannel.ts

```diff
@@ -0,0 +1,279 @@
+ * X25519/ECDH key agreement based secure rendezvous channel.
+ * Note that this is UNSTABLE and may have breaking changes without notice.
+ */
+export class MSC3903ECDHv1RendezvousChannel implements RendezvousChannel {
+    private olmSAS?: SAS;
+    private ourPublicKey: Uint8Array;
+    private aesKey?: CryptoKey | Uint8Array;
+
+    public constructor(
+        public transport: RendezvousTransport,
+        private theirPublicKey?: Uint8Array,
+        public onFailure?: (reason: RendezvousFailureReason) => void,
+    ) {
+        this.olmSAS = new global.Olm.SAS();
+        this.ourPublicKey = decodeBase64(this.olmSAS.get_pubkey());
+    }
+
+    public async generateCode(intent: RendezvousIntent): Promise<ECDHv1RendezvousCode> {
+        if (this.transport.ready) {
+            throw new Error('Code already generated');
+        }
+
+        await this.send({ algorithm: RendezvousChannelAlgorithm.ECDH_V1 });
+
+        const rendezvous: ECDHv1RendezvousCode = {
+            "rendezvous": {
+                algorithm: RendezvousChannelAlgorithm.ECDH_V1,
+                key: encodeBase64(this.ourPublicKey),
+                transport: await this.transport.details(),
+            },
+            intent,
+        };
+
+        return rendezvous;
+    }
+
+    public async connect(): Promise<string> {
+        if (!this.olmSAS) {
+            throw new Error('Channel closed');
+        }
+
+        const isInitiator = !this.theirPublicKey;
+
+        if (isInitiator) {
+            // wait for the other side to send us their public key
+            const res = await this.receive();
+            if (!res) {
+                throw new Error('No response from other device');
+            }
+            const { key, algorithm } = res;
+
+            if (algorithm !== RendezvousChannelAlgorithm.ECDH_V1 || !key) {
+                throw new RendezvousError(
+                    'Unsupported algorithm: ' + algorithm,
+                    RendezvousFailureReason.UnsupportedAlgorithm,
+                );
+            }
+
+            this.theirPublicKey = decodeBase64(key);
+        } else {
+            // send our public key unencrypted
+            await this.send({
+                algorithm: RendezvousChannelAlgorithm.ECDH_V1,
+                key: encodeBase64(this.ourPublicKey),
+            });
+        }
+
+        this.olmSAS.set_their_key(encodeBase64(this.theirPublicKey!));
+
+        const initiatorKey = isInitiator ? this.ourPublicKey : this.theirPublicKey!;
+        const recipientKey = isInitiator ? this.theirPublicKey! : this.ourPublicKey;
+        let aesInfo = RendezvousChannelAlgorithm.ECDH_V1.toString();
+        aesInfo += `|${encodeBase64(initiatorKey)}`;
+        aesInfo += `|${encodeBase64(recipientKey)}`;
+
+        const aesKeyBytes = this.olmSAS.generate_bytes(aesInfo, 32);
+
+        this.aesKey = await importKey(aesKeyBytes);
```

Comment:
(After you switch things over to WebCrypto, ) you should clear out `aesKeyBytes` by [`fill`-ing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/fill) it with 0's, to make sure that the key doesn't remain in memory..